File: sdk/go/provider/inboxkitten.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/url"
	"regexp"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const inboxkittenBase = "https://inboxkitten.com/api/v1/mail"
const inboxkittenDomain = "inboxkitten.com"

type inboxkittenListItem struct {
	Storage struct {
		Region string `json:"region"`
		Key    string `json:"key"`
	} `json:"storage"`
	Message struct {
		Headers struct {
			From    string `json:"from"`
			Subject string `json:"subject"`
		} `json:"headers"`
	} `json:"message"`
	Envelope struct {
		Sender  string `json:"sender"`
		Targets string `json:"targets"`
	} `json:"envelope"`
	Recipient string  `json:"recipient"`
	Timestamp float64 `json:"timestamp"`
}

type inboxkittenMeta struct {
	Name       string `json:"name"`
	Subject    string `json:"subject"`
	Recipients string `json:"recipients"`
}

func inboxkittenLocal() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	var b strings.Builder
	b.WriteString("sdk")
	for i := 0; i < 16; i++ {
		b.WriteByte(chars[rand.Intn(len(chars))])
	}
	return b.String()
}

func inboxkittenGet(u string, accept string) ([]byte, error) {
	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", accept)
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("inboxkitten: http %d", resp.StatusCode)
	}
	return body, nil
}

func inboxkittenHTMLText(html string) string {
	reScript := regexp.MustCompile(`(?is)<script[\s\S]*?</script>`)
	reTag := regexp.MustCompile(`(?s)<[^>]+>`)
	text := reScript.ReplaceAllString(html, " ")
	text = reTag.ReplaceAllString(text, " ")
	text = strings.NewReplacer(
		"&nbsp;", " ",
		"&amp;", "&",
		"&lt;", "<",
		"&gt;", ">",
		"&quot;", `"`,
		"&#39;", "'",
	).Replace(text)
	return strings.Join(strings.Fields(text), " ")
}

func InboxkittenGenerate() (*CreatedMailbox, error) {
	local := inboxkittenLocal()
	return &CreatedMailbox{Channel: "inboxkitten", Email: local + "@" + inboxkittenDomain}, nil
}

func inboxkittenFetchDetail(row inboxkittenListItem, recipient string) map[string]any {
	raw := map[string]any{
		"id":        row.Storage.Key,
		"messageId": row.Storage.Key,
		"from":      row.Message.Headers.From,
		"to":        recipient,
		"subject":   row.Message.Headers.Subject,
		"timestamp": row.Timestamp,
	}
	if raw["from"] == "" {
		raw["from"] = row.Envelope.Sender
	}
	if row.Recipient != "" {
		raw["to"] = row.Recipient
	} else if row.Envelope.Targets != "" {
		raw["to"] = row.Envelope.Targets
	}
	if row.Storage.Region == "" || row.Storage.Key == "" {
		return raw
	}

	q := url.Values{}
	q.Set("region", row.Storage.Region)
	q.Set("key", row.Storage.Key)
	metaBody, metaErr := inboxkittenGet(inboxkittenBase+"/getKey?"+q.Encode(), "application/json")
	htmlBody, htmlErr := inboxkittenGet(inboxkittenBase+"/getHtml?"+q.Encode(), "text/html,*/*")
	if metaErr == nil {
		var meta inboxkittenMeta
		if json.Unmarshal(metaBody, &meta) == nil {
			if meta.Name != "" {
				raw["from"] = meta.Name
			}
			if meta.Recipients != "" {
				raw["to"] = meta.Recipients
			}
			if meta.Subject != "" {
				raw["subject"] = meta.Subject
			}
		}
	}
	if htmlErr == nil {
		html := string(htmlBody)
		raw["html"] = html
		raw["text"] = inboxkittenHTMLText(html)
	}
	return raw
}

func InboxkittenGetEmails(email string) ([]NormEmail, error) {
	parts := strings.Split(strings.TrimSpace(email), "@")
	local := parts[0]
	if local == "" {
		return nil, fmt.Errorf("inboxkitten: empty email")
	}

	q := url.Values{}
	q.Set("recipient", local)
	body, err := inboxkittenGet(inboxkittenBase+"/list?"+q.Encode(), "application/json")
	if err != nil {
		return nil, err
	}
	var rows []inboxkittenListItem
	if err := json.Unmarshal(body, &rows); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		out = append(out, NormalizeMap(inboxkittenFetchDetail(row, email), email))
	}
	return out, nil
}

File: sdk/npm/src/providers/inboxkitten.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'inboxkitten';
const API_BASE = 'https://inboxkitten.com/api/v1/mail';
const DOMAIN = 'inboxkitten.com';

interface ListItem {
  storage?: {
    region?: string;
    key?: string;
  };
  message?: {
    headers?: {
      from?: string;
      subject?: string;
    };
  };
  envelope?: {
    sender?: string;
    targets?: string;
  };
  recipient?: string;
  timestamp?: number;
}

interface MessageMeta {
  name?: string;
  subject?: string;
  recipients?: string;
}

function randomLocal(): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let out = 'sdk';
  for (let i = 0; i < 16; i++) out += chars[Math.floor(Math.random() * chars.length)];
  return out;
}

function decodeHtmlEntities(s: string): string {
  return s
    .replace(/&#x([0-9a-fA-F]+);/g, (_, h) => String.fromCodePoint(parseInt(h, 16)))
    .replace(/&#(\d+);/g, (_, d) => String.fromCodePoint(parseInt(d, 10)))
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

function htmlToText(html: string): string {
  return decodeHtmlEntities(html.replace(/<script[\s\S]*?<\/script>/gi, ' ').replace(/<[^>]+>/g, ' '))
    .replace(/\s+/g, ' ')
    .trim();
}

function localPart(email: string): string {
  return String(email || '').trim().split('@')[0] || '';
}

async function fetchJSON<T>(url: string): Promise<T> {
  const response = await fetchWithTimeout(url, {
    method: 'GET',
    headers: { Accept: 'application/json' },
  });
  if (!response.ok) {
    throw new Error(`inboxkitten http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

async function fetchHtml(region: string, key: string): Promise<string> {
  const url = `${API_BASE}/getHtml?region=${encodeURIComponent(region)}&key=${encodeURIComponent(key)}`;
  const response = await fetchWithTimeout(url, {
    method: 'GET',
    headers: { Accept: 'text/html,*/*' },
  });
  if (!response.ok) {
    throw new Error(`inboxkitten html ${response.status}`);
  }
  return response.text();
}

async function fetchMeta(region: string, key: string): Promise<MessageMeta> {
  const url = `${API_BASE}/getKey?region=${encodeURIComponent(region)}&key=${encodeURIComponent(key)}`;
  return fetchJSON<MessageMeta>(url);
}

async function fetchDetail(row: ListItem, recipient: string): Promise<Record<string, unknown>> {
  const region = String(row.storage?.region || '').trim();
  const key = String(row.storage?.key || '').trim();
  const base = {
    id: key,
    messageId: key,
    from: row.message?.headers?.from || row.envelope?.sender || '',
    to: row.recipient || row.envelope?.targets || recipient,
    subject: row.message?.headers?.subject || '',
    timestamp: row.timestamp,
  };
  if (!region || !key) return base;

  const [meta, html] = await Promise.all([fetchMeta(region, key), fetchHtml(region, key)]);
  return {
    ...base,
    from: meta.name || base.from,
    to: meta.recipients || base.to,
    subject: meta.subject || base.subject,
    text: htmlToText(html),
    html,
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const local = randomLocal();
  return { channel: CHANNEL, email: `${local}@${DOMAIN}` };
}

export async function getEmails(email: string): Promise<Email[]> {
  const local = localPart(email);
  if (!local) throw new Error('inboxkitten: empty email');

  const rows = await fetchJSON<ListItem[]>(`${API_BASE}/list?recipient=${encodeURIComponent(local)}`);
  if (!Array.isArray(rows)) return [];

  const emails: Email[] = [];
  for (const row of rows) {
    try {
      emails.push(normalizeEmail(await fetchDetail(row, email), email));
    } catch {
      emails.push(normalizeEmail({
        id: row.storage?.key || '',
        from: row.message?.headers?.from || row.envelope?.sender || '',
        to: row.recipient || row.envelope?.targets || email,
        subject: row.message?.headers?.subject || '',
        timestamp: row.timestamp,
      }, email));
    }
  }
  return emails;
}
